### DISCUSSION TITLE
Epoptes crashes when a user logs off

### DISCUSSION THREAD

**Discussion by frankzimper** [epoptes/epoptes#33] [Q&A]
I run epoptes from a fat client using ltsp-remoteapps - so it actually runs on the server. When a client logs off the epoptes GUI closes.

[epoptes.log](https://github.com/Epoptes/epoptes/files/1345713/epoptes.log)

**Comment by maxnet** [comment]
Likely related to: https://github.com/Epoptes/epoptes/pull/41

Caused by Epoptes assuming a notification daemon is always running.
(although the Debian package pulls in notification-daemon, that is not started automatically)

**Comment by frankzimper** [comment]
Thanks! 
I implemented this fix and the problem is gone.

**Comment by alkisg** [comment]
The older 0.5.10 Epoptes version is based on Python2/Gtk2 and will not receive any updates.

I uploaded the newer 1.0.0 version in the [Epoptes proposed PPA](http://www.epoptes.org/documentation/ppa) for Ubuntu 18.10, 18.04 and 16.04.
If you still have a the problem with that version, I could take a look.

**Comment by alkisg** [comment]
No reply in 2 months, closing this issue.

**Comment by NdK73** [comment]
I still see server crashes when powering off clients.
Debian 10 Buster freshly updated (epoptes/stable,now 1.0.1-2).
notification-daemon is installed (default settings).

**Comment by alkisg** [comment]
Could you run `epoptes` from a terminal, reproduce the crash, and paste the terminal output?

Also, which desktop environment?

**Comment by NdK73** [comment]
No desktop environment: just ssh -X to the (headless) server. I think it should be quite a common use.
The whole session is:
```
$ epoptes
Gtk-Message: 09:16:29.682: Failed to load module "canberra-gtk-module"
Got clients: None
Sending magic packet to 54B2037006AB
New connection from 10.10.10.180:56600
New connection from 10.10.10.180:56598
add_client's been called for 10.10.10.180:56598
  Old client: hostname=str957-i-80, type=standalone, uid=0, user=root
add_client's been called for 10.10.10.180:56600
  Old client: hostname=str957-i-80, type=standalone, uid=0, user=root
Disconnect from 10.10.10.180:56600
Unhandled Error
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/twisted/protocols/amp.py", line 1043, in _commandReceived
    deferred = self.dispatchCommand(box)
  File "/usr/lib/python3/dist-packages/twisted/protocols/amp.py", line 1101, in dispatchCommand
    return maybeDeferred(responder, box)
  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 151, in maybeDeferred
    result = f(*args, **kw)
  File "/usr/lib/python3/dist-packages/twisted/protocols/amp.py", line 1188, in doit
    return maybeDeferred(aCallable, **kw).addCallback(
--- <exception caught here> ---
  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 654, in _runCallbacks
    current.result = callback(current.result, *args, **kw)
  File "/usr/lib/python3/dist-packages/twisted/protocols/amp.py", line 1171, in checkKnownErrors
    key = error.trap(*command.allErrors)
  File "/usr/lib/python3/dist-packages/twisted/python/failure.py", line 439, in trap
    self.raiseException()
  File "/usr/lib/python3/dist-packages/twisted/python/failure.py", line 467, in raiseException
    raise self.value.with_traceback(self.tb)
  File "/usr/lib/python3/dist-packages/twisted/internet/defer.py", line 151, in maybeDeferred
    result = f(*args, **kw)
  File "/usr/lib/python3/dist-packages/epoptes/daemon/uiconnection.py", line 43, in client_disconnected
    self.client.amp_client_disconnected(handle)
  File "/usr/lib/python3/dist-packages/epoptes/ui/gui.py", line 728, in amp_client_disconnected
    _("Shut down:"), client.get_name())
  File "/usr/lib/python3/dist-packages/epoptes/ui/notifications.py", line 102, in enqueue
    self.show()
  File "/usr/lib/python3/dist-packages/epoptes/ui/notifications.py", line 135, in show
    self.notification.show()
gi.repository.GLib.Error: g-dbus-error-quark: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.Notifications was not provided by any .service files (2)
```
The shutdown was sent just before `Disconnect from 10.10.10.180:56600`.

**Comment by alkisg** [comment]
`ssh -X` doesn't forward notifications; you need dbus-launch for that. I.e. try:

```shell
ssh -X server
dbus-launch epoptes
```

Related documentation: http://www.epoptes.org/documentation/fat-clients

**Comment by NdK73** [comment]
Installed `dbus-x11` to have dbus-launch, then
```$ dbus-launch epoptes
Gtk-Message: 09:49:27.390: Failed to load module "canberra-gtk-module"

(epoptes:6617): Gdk-ERROR **: 09:49:27.528: The program 'epoptes' received an X Window System error.
This probably reflects a bug in the program.
The error was 'BadAccess (attempt to access private resource denied)'.
  (Details: serial 159 error_code 10 request_code 130 (MIT-SHM) minor_code 1)
  (Note to programmers: normally, X errors are reported asynchronously;
   that is, you will receive the error a while after causing it.
   To debug your program, run it with the GDK_SYNCHRONIZE environment
   variable to change this behavior. You can then get a meaningful
   backtrace from your debugger if you break on the gdk_x_error() function.)
Rilevato trace/breakpoint
```
I'm accessing the server from a management machine outside the lab. And the lab is not using LTSP.

Edit: reconnected using `ssh -Y server` and dbus-launch ran epoptes, but at shutdown the same dbus error appears.

**Comment by alkisg** [comment]
There are probably some more packages missing for notification-daemon to work properly without a desktop environment.
Epoptes depends on notification-daemon.
Notification-daemon should depend on whatever else it needs. It appears that it doesn't; solving the correct notification-daemon dependencies isn't an epoptes issue though.

What the epoptes code could do, would be "to work even if notification-daemon fails", but this isn't a priority, as notification-daemon usually works even over ssh -X; in your case it's just missing some packages because it's headless, which isn't a frequent use case.

You may file a bug report against whatever notification-daemon you're using...
